port/linux/tcpsession.c

```diff
@@ -0,0 +1,1214 @@
+      tcp_send_waiting_messages_locked(ws);
+      tcp_connect_add_data(&ws->endpoint, OC_TCP_SOCKET_STATE_CONNECTED,
+                           ws->on_tcp_connect, ws->on_tcp_connect_data);
+      oc_memb_free(&g_tcp_waiting_session_s, ws);
+      return OC_TCP_SOCKET_STATE_CONNECTED;
+    }
+    OC_ERR("cannot allocate ongoing TCP connection");
+  }
+
+  ws->retry.count++;
+  ws->retry.start = now;
+  ws->sock = cs.socket;
+  tcp_waiting_session_set_socked_locked(&ws->dev->tcp, ws->sock);
+  return cs.state;
+}
+
+oc_clock_time_t
+tcp_check_expiring_sessions(oc_clock_time_t now)
+{
+  oc_clock_time_t expires_in = 0;
+  pthread_mutex_lock(&g_mutex);
+  for (tcp_waiting_session_t *ws =
+         (tcp_waiting_session_t *)oc_list_head(g_waiting_session_list);
+       ws != NULL;) {
+    tcp_waiting_session_t *ws_next = ws->next;
+    int check = tcp_waiting_session_check(ws, now);
+    switch (check) {
+    case TCP_WAITING_SESSION_RETRY: {
+      if (tcp_retry_waiting_session_locked(ws, now) !=
+          OC_TCP_SOCKET_STATE_CONNECTED) {
+        expires_in = g_connect_retry.timeout * OC_CLOCK_SECOND;
+      }
+      break;
+    }
+    case TCP_WAITING_SESSION_VALID: {
+      oc_clock_time_t ei =
+        (g_connect_retry.timeout * OC_CLOCK_SECOND) - (now - ws->retry.start);
+      if (expires_in == 0 || ei < expires_in) {
+        expires_in = ei;
+      }
+      break;
+    }
+    case TCP_WAITING_SESSION_EXPIRED:
+      free_waiting_session_locked(ws, true, false);
+      break;
+    default:
+      break;
+    }
+    ws = ws_next;
+  }
+  pthread_mutex_unlock(&g_mutex);
+  tcp_connect_process_data();
+  return expires_in;
+}
+
+int
+tcp_process_waiting_sessions(fd_set *fds)
+{
+  int ret = 0;
+  tcp_on_connect_data_t on_connect = {
+    .state = -1,
+    .fn = NULL,
+    .fn_data = NULL,
+  };
+  pthread_mutex_lock(&g_mutex);
+  for (tcp_waiting_session_t *ws =
+         (tcp_waiting_session_t *)oc_list_head(g_waiting_session_list);
+       ws != NULL; ws = ws->next) {
+    if (!FD_ISSET(ws->sock, fds)) {
+      continue;
+    }
+
+    OC_DBG("tcp session(%p) connect (fd=%d): %u", ws, ws->sock,
+           (unsigned)ws->retry.count);
+    FD_CLR(ws->sock, fds);
+    ret = 1;
+    if (tcp_try_connect_waiting_session_locked(ws)) {
+      tcp_send_waiting_messages_locked(ws);
+      if (ws->on_tcp_connect != NULL) {
+        memcpy(&on_connect.endpoint, &ws->endpoint, sizeof(oc_endpoint_t));
+        on_connect.state = OC_TCP_SOCKET_STATE_CONNECTED;
+        on_connect.fn = ws->on_tcp_connect;
+        on_connect.fn_data = ws->on_tcp_connect_data;
+      }
+      oc_memb_free(&g_tcp_waiting_session_s, ws);
+    } else {
+      OC_DBG("failed to connect session(%p)", ws);
+      tcp_context_cfds_fd_clr(&ws->dev->tcp, ws->sock);
+      close(ws->sock);
+      ws->sock = -1;
+    }
+    break;
+  }
+  pthread_mutex_unlock(&g_mutex);
+  if (on_connect.fn != NULL) {
+    on_connect.fn(&on_connect.endpoint, on_connect.state, on_connect.fn_data);
+  }
```

Comment:
`on_connect.fn` is called from the network thread and not from event thread - user doesn't expect that